### ISSUE TITLE
`uv pip compile --resolution lowest-direct` treats transitive dependencies of file dependencies as direct.

### ISSUE THREAD

**Issue by patrick-kidger** [astral-sh/uv#4238]
After a bit of head-scratching, here's a WME:
```
# one/pyproject.toml
[project]
name = "one"
version = "0.0.1"
dependencies = ["two @ file://${PROJECT_ROOT}/../two"]

# two/pyproject.toml
[project]
name = "two"
version = "0.0.1"
dependencies = ["typing_extensions>=4.5.0"]
```
then
```
one❯ uv pip compile pyproject.toml --resolution lowest-direct                
Resolved 2 packages in 86ms
# This file was autogenerated by uv via the following command:
#    uv pip compile pyproject.toml --resolution lowest-direct
two @ file://${PROJECT_ROOT}/../two
    # via one (pyproject.toml)
typing-extensions==4.5.0
    # via two
```

The same bug also occurs in all of the following scenarios:
1. When expressed on the command line:
```
one❯ printf 'two @ file://${PROJECT_ROOT}/../two' | uv pip compile - --resolution lowest-direct
```
2. When expressed editably (side note, this one is our actual use-case).
```
one❯ printf '-e file://${PROJECT_ROOT}/../two' | uv pip compile - --resolution lowest-direct
```
3. When not using `PROJECT_ROOT`:
```
one❯ printf '-e file:///absolute/path/to/two' | uv pip compile - --resolution lowest-direct
```

(FWIW we found this problem as the resulting `requirements.txt` files go on to be used as `constraints` in one of our later build steps.)

**Comment by notatallshaw-gts** [comment]
I also encountered this same behavior yesterday, specifically in the context of "3.", and while surprised I personally found it useful, so wasn't sure if it was intentional or not.

If I have a local package I am passing into the requirement, uv can only see one version of it, so it made sense to me it would look for the lowest direct dependencies of that package, and was a better test for my environment.

But it is surprising, and not the behavior I would expect from the description.

**Comment by charliermarsh** [comment]
Yeah this is intentional -- we consider anything local to be first-party, i.e., conceptually part of your local workspace.

**Comment by charliermarsh** [comment]
Is this a real problem or undesirable behavior in your use-case, or just something that needs to be better-documented? (I'm not fully convinced that we should change it but open-minded of course.)

**Comment by patrick-kidger** [comment]
At least for us it's undesirable behaviour! What's going on is that we have several libraries that depend on each other. Most of the time we express this as a git dependency. But during local development, if making a cross-cutting change, then we switch that into an editable dependency. So
(a) ideally such dependencies would behave the same as normal PyPI dependencies, as they're still conceptually separate libraries;
(b) if I've grokked this correctly then I think the nature of our dependency resolution actually changes depending on whether we run `uv compile` whilst the editable dependency is in place or not!

**Comment by charliermarsh** [comment]
Yeah, if you run with `--resolution lowest-direct`, then what's considered "direct" will change if you switch from using a Git dependency to a local dependency, because one is considered direct and the other is not.

I mean, the weird thing is, given:

```
# one/pyproject.toml
[project]
name = "one"
version = "0.0.1"
dependencies = ["typing-extensions"]
```

And then `uv pip install ./one`, `typing-extensions` isn't technically a direct dependency either -- it's a transitive dependency, the direct dependency is `one @ file:///path/to/tone`. But intuitively it should be treated as a direct dependency, right?  Similarly, given `uv pip install ./one` vs. `uv pip install "one @ git+https://...`, we treat `typing-extensions` as direct in the first case but not in the second case, so switching the source does again change what's considered direct. Separate from whether this should be applied transitively as it is now, do you think both of those decisions are correct?

**Comment by patrick-kidger** [comment]
> But intuitively it should be treated as a direct dependency, right?

I don't think so -- at least subjectively by own intuition! :D

Like you say, these are transitive dependencies, I don't have a mental model that differs from that.

> Similarly, given `uv pip install ./one` vs. `uv pip install "one @ git+https://...`, we treat typing-extensions as direct in the first case but not in the second case, so switching the source does again change what's considered direct.

Likewise, this doesn't fit my intution either. I don't expect where a dependency is located to affect anything whatsoever. (Beyond the possible need to make a network request.)

**Comment by notatallshaw-gts** [comment]
I would say that yesterday I arrived at the same mental model as @charliermarsh.

The local projects are projects that I own, so while these local project direct dependencies are transitive of the requirements I feed to uv, they are still the direct dependencies of _my_ projects, which is more useful to my use case at least.

CC @potiuk, is airflow's CI use of lowest-direct referring to local files? A change in behavior here could break your setup.
